scijava/scijava-ops-engine/src/test/java/org/scijava/ops/engine/yaml/ops/YAMLClassOp.java

```diff
@@ -7,6 +7,10 @@
 
 /**
  * An example Op, registered by YAML into the {@link OpEnvironment}
+ * @implNote op names=example.add
+ * @input aDouble the first input
+ * @input bDouble the second input
+ * @output the result
```

Comment:
Man, putting the implementation details here feels like a vulnerability to action-at-a-distance errors - compilers and/or IDE tools won't see the correlation between this and (a) the Javadoc on the functional method and (b) the functional method parameters.

But, at the same time, keeping  the YAML parser general means probably not special-casing logic to pull tags from functional methods.

If you write an Op as a class, you probably don't want to copy your javadoc up to the class, right?

I don't know, does anyone else have opinions? @ctrueden @hinerm

Reply:
Solution: discontinue support for class-based op implementations! :clown_face:

Reply:
In seriousness: I assume there are reasons why class-based op implementations are sometimes necessary/preferred over method-based ones? (I know for field-based ones, we cannot support e.g. op dependencies.)

Reply:
My half-baked opinion is: how about changing class-based op implementations to read that stuff from the functional method, and *not* from the class? The only weirdness there is that people will need to put javadoc on an `@Override`n method, but that's certainly a permitted thing to do, and would then be documenting the actual function.

Reply:
> In seriousness: I assume there are reasons why class-based op implementations are sometimes necessary/preferred over method-based ones? (I know for field-based ones, we cannot support e.g. op dependencies.)

I don't think there's any reason why it's necessary, but the preference thing I think is important. I do not want to make [`RichardsonLucyC`](https://github.com/scijava/incubator/blob/105acb4fd5aaccac673ef7ed23f9b6b5acba17e9/imagej/imagej-ops2/src/main/java/net/imagej/ops2/deconvolve/RichardsonLucyC.java#L62) a `Method` - imagine how unwieldy that'd be as a single method :exploding_head:

Reply:
> My half-baked opinion is: how about changing class-based op implementations to read that stuff from the functional method, and _not_ from the class? The only weirdness there is that people will need to put javadoc on an `@Override`n method, but that's certainly a permitted thing to do, and would then be documenting the actual function.

That's how it was done with therapi, and I certainly think it is the nicer way from the Op author's perspective to do things. However I don't think it's general enough beyond Ops - what if we start using YAML to expose other types of implementations? Is that something we'd want to do?

Here's what we 

> My half-baked opinion is: how about changing class-based op implementations to read that stuff from the functional method, and _not_ from the class? The only weirdness there is that people will need to put javadoc on an `@Override`n method, but that's certainly a permitted thing to do, and would then be documenting the actual function.

Well, this is what was done with therapi, and it is imo the nicest way to do things. And we could make a specialization on FunctionalInterfaces that pulls javadoc tags from *both* the class and the functional method - the only thing to decide then is where the logic lives.

Right now, there's probably sufficient logic within SciJava Types for that, so the question is whether SciJava Javadoc Parser should depend on SciJava Types. That'd make the build even more of a nightmare, so I'm tempted to just fork the necessary logic into this class...

Reply:
I don't understand why the scijava javadoc parser needs to reason about geneic types?

Reply:
The SciJava Javadoc Parser is more specialized, only looking for and storing elements with `@implNote` implementations. In the case of `Class` elements, the `@implNote` will be on the class. Thus, to pull any tags off of the functional method, we'll have to:

1) Determine the functional interface implemented by the Class
2) Determine the functional method implemented
3) Find that method and grab all tags.

The first two steps can be done through SciJava Types.

Reply:
In order to support method-based op declarations, the SciJava javadoc parser is already responding to implNote declarations on methods, right? Can we not simply use the already-recorded implNote details on the method, without doing all the work you describe? In other words: the burden of understanding how to interpret the implNote and related indexed metadata would remain within the Ops framework, not the javadoc parser? So then it's just a matter of disambiguating the two different kinds of methods that will have implNotes: either an op-method kind, or a functional-interface-method-of-op-class kind?

Reply:
Another way to articulate my concern here: the javadoc parser is writing out an Ops-style YAML file directly, right? If so, it's not really the "scijava javadoc parser" at that point, but rather the *Ops* javadoc parser. No? Or... does the scijava javadoc parser write out some other intermediate thing? How exactly does it work?

Reply:
> So then it's just a matter of disambiguating the two different kinds of methods that will have implNotes: either an op-method kind, or a functional-interface-method-of-op-class kind?

Hmm, yeah, that could work. Maybe I'll give it a shot.

> Another way to articulate my concern here: the javadoc parser is writing out an Ops-style YAML file directly, right? If so, it's not really the "scijava javadoc parser" at that point, but rather the Ops javadoc parser. No? Or... does the scijava javadoc parser write out some other intermediate thing? How exactly does it work?

It's more general than ops, for each `@implNote blah`, a file `blah.yaml` is output in the target folder containing all elements (classes, fields, methods, may also want to consider Java 16 records) with that tag. I explain it pretty succinctly and with an explanation in the PR description up top.

Reply:
Although, now that I look at it, I think the example yaml I put in the PR description is a little different than what is recorded now...

Reply:
One other thing I'll note: I am really interested in others' opinions on whether this annotation processor *would* be useful in contexts beyond ops - if you think it is not, then yeah, we should probably rename it and specialize a bit for Ops.

Reply:
We no longer need to put the tags on the class!